### PR TITLE
#2305 Renamed DynHaloExchange to LFRicHaloExchange

### LINKED ISSUE #2305: DynHaloExchange to LFRicHaloExchange
sub-issue of #2235

### PR DESCRIPTION
**Pull request by oakleybrunt**
Potentially closes #2305 which is a sub-issue of #2235

### CODE REVIEW

File: src/psyclone/tests/dynamo0p3_test.py [hunk 1]
```diff
@@ -2872,7 +2872,7 @@ def test_multiple_derived_type_args(dist_mem, tmpdir):
 
 def test_haloexchange_unknown_halo_depth():
     ''' If a stencil extent is provided in the kernel metadata then the
-    value is stored in an instance of the DynHaloExchange class. This test
+    value is stored in an instance of the LFRicHaloExchange class. This test
```
Comment (TeranIvy) [inline src/psyclone/tests/dynamo0p3_test.py:2875]: Please add yourself under "Modified".

File: src/psyclone/tests/psyGen_test.py [hunk 2]
```diff
@@ -1551,7 +1551,7 @@ def test_haloexchange_node_str():
     invoke = psy.invokes.invoke_list[0]
     schedule = invoke.schedule
     # We have to manually call the correct node_str() method as the one we want
-    # to test is overridden in DynHaloExchange.
+    # to test is overridden in LFRicHaloExchange.
```
Comment (TeranIvy) [inline src/psyclone/tests/psyGen_test.py:1554]: Please add yourself under "Modified".

File: src/psyclone/tests/psyGen_test.py [hunk 3]
```diff
@@ -1551,7 +1551,7 @@ def test_haloexchange_node_str():
     invoke = psy.invokes.invoke_list[0]
     schedule = invoke.schedule
```
Comment (TeranIvy) [inline src/psyclone/tests/psyGen_test.py:1552]: Not the fault of this PR, but `pylint` reports
```
psyGen_test.py:67:0: W0611: Unused CWriter imported from psyclone.psyir.backend.c (unused-import)
psyGen_test.py:68:0: W0611: Unused FortranWriter imported from psyclone.psyir.backend.fortran (unused-import)
```
Would you mind improving this?

File: tutorial/practicals/LFRic/distributed_memory/3_overlapping_comms/solutions/overlap_generic.py [hunk 4]
```diff
@@ -76,7 +76,7 @@ def trans(psy):
         # Get the schedule (the PSyIR representation of the PSy-layer)
         schedule = invoke.schedule
         # Split any synchronous halo exchanges into asynchronous halo exchanges
-        for hex_node in schedule.walk(DynHaloExchange):
+        for hex_node in schedule.walk(LFRicHaloExchange):
```
Comment (TeranIvy) [inline tutorial/practicals/LFRic/distributed_memory/3_overlapping_comms/solutions/overlap_generic.py:79]: Please add yourself under "Modified", i.e.
```
# Author: R. W. Ford, STFC Daresbury Lab
# Modified: O. Brunt, Met Office
```

Please also update the modification year to 2023.

**Review by TeranIvy — CHANGES_REQUESTED** [review]
Nice work!

I would suggest going all the way and renaming all DynHaloExchange* classes in this PR. Splitting of classes can wait for another PR, as it would muddle waters.

The remarks I have are minor and are mainly about code style and authorship. Speaking of authorship, modifying developer guide includes adding the developer's name to the list of developers' guide authors. Please see inline comment on how to do that.

The branch will also need to be updated with the master before coming back for the second review.

File: src/psyclone/dynamo0p3.py [hunk 5]
```diff
@@ -5793,7 +5793,7 @@ def gen_code(self, parent):
         parent.add(CommentGen(parent, ""))
 
 
-class DynHaloExchangeStart(DynHaloExchange):
+class DynHaloExchangeStart(LFRicHaloExchange):
```
[the remark this answers is not in the document]
Comment (oakleybrunt) [reply src/psyclone/dynamo0p3.py]: Renamed all DynHaloExchange* to their LFRicHaloExchange* version

File: src/psyclone/tests/dynamo0p3_multigrid_test.py [hunk 6]
```diff
@@ -50,7 +50,7 @@
 from fparser import api as fpapi
 from psyclone.configuration import Config
 from psyclone.domain.lfric import LFRicConstants
-from psyclone.dynamo0p3 import DynHaloExchange, DynKernMetadata, HaloReadAccess
+from psyclone.dynamo0p3 import LFRicHaloExchange, DynKernMetadata, HaloReadAccess
```
[the remark this answers is not in the document]
Comment (oakleybrunt) [reply src/psyclone/tests/dynamo0p3_multigrid_test.py:53]: You accidentally quoted the wrong `pycodestyle` message. Putting the correct one here for reference but thank you for pointing this out. 

`src/psyclone/tests/dynamo0p3_multigrid_test.py:53:80: E501 line too long (81 > 79 characters)`

File: src/psyclone/dynamo0p3.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/dynamo0p3.py]: Thanks :)

File: doc/developer_guide/conf.py [hunk 7]
```diff
@@ -247,8 +247,8 @@
 #  documentclass [howto/manual]).
 latex_documents = [
     ('index', 'psyclone-dev.tex', u"PSyclone Developer Guide",
-     u'Rupert Ford, Joerg Henrichs, Iva Kavcic, Andrew Porter,\\\\ Sergi Siso'
-     ' and Lottie Turner', 'manual'),
+     u'Oakley Brunt, Rupert Ford, Joerg Henrichs, Iva Kavcic, \\\\'
```
Comment (TeranIvy) [inline doc/developer_guide/conf.py:250]: Thanks for adding your name :) There are a few more documentation options where it needs to be updated: `man_pages`, `texinfo_documents`, `epub_author` and `epub_publisher` (pretty much where you see Lottie's and my name.

File: examples/lfric/scripts/async_halo_exchanges.py [hunk 8]
```diff
@@ -32,6 +32,7 @@
 # POSSIBILITY OF SUCH DAMAGE.
 # -----------------------------------------------------------------------------
 # Authors: R. Ford, A. R. Porter and S. Siso, STFC Daresbury Laboratory
+# Modified by O. Brunt, Met Office
```
Comment (TeranIvy) [inline examples/lfric/scripts/async_halo_exchanges.py:35]: I would suggest `# Modified: O. Brunt, Met Office` for consistency with the "Authors" line above.

File: src/psyclone/tests/psyGen_test.py [hunk 9]
```diff
@@ -32,7 +32,7 @@
 # POSSIBILITY OF SUCH DAMAGE.
 # -----------------------------------------------------------------------------
 # Authors: R. W. Ford, A. R. Porter, S. Siso and N. Nobre, STFC Daresbury Lab
-# Modified I. Kavcic and L. Turner, Met Office
+# Modified I. Kavcic, L. Turner and O. Brunt, Met Office
```
Comment (TeranIvy) [inline src/psyclone/tests/psyGen_test.py:35]: I would suggest adding `:` after `# Modified` for consistency with the "Authors" line above.

File: tutorial/practicals/LFRic/distributed_memory/3_overlapping_comms/solutions/overlap_generic.py [hunk 10]
```diff
@@ -32,6 +32,7 @@
 # POSSIBILITY OF SUCH DAMAGE.
 # -----------------------------------------------------------------------------
 # Author: R. W. Ford, STFC Daresbury Lab
+# Modified by O. Brunt, Met Office
```
Comment (TeranIvy) [inline tutorial/practicals/LFRic/distributed_memory/3_overlapping_comms/solutions/overlap_generic.py:35]: I would suggest `# Modified: O. Brunt, Met Office` for consistency with the "Authors" line above.

**Review by TeranIvy — CHANGES_REQUESTED** [review]
Almost there :) There are a few minor things to address, as well as a request to rename and move one of the tests. The branch also needs an update with the latest master.

**Review by TeranIvy — APPROVED** [review]
All well now. Updated changelog and user guide. CI passes. Tests with compilation pass. Proceeding to merge.

**Pull request merged**